gadgets/trace_exec/program.bpf.c

```diff
@@ -38,9 +39,13 @@ struct event {
 	bool pupper_layer;
 	unsigned int args_size;
 	char comm[TASK_COMM_LEN];
+	char cwd[MAX_STRING_SIZE];
```

Comment:
@mauriciovasquezbernal, @blanquicet, any suggestions on how to avoid redundant declaration of `cwd` for when --paths=false?

Reply:
In my opinion, the semantics of `macro: WITH_PATHS` are not good: your yaml snippet is presumably gadget.yaml, which is run by `ig run` to load the gadget correctly. But C macros are used at compiled time and I don't think that compilation details should leak in gadget.yaml. The gadget might be built from other languages than C and it should not matter for `ig run`. Maybe it would work better in build.yaml?

But I am still worried about the combinatorial explosion if several fields use such macros.

I think we should find a way to pass the type in the ring buffer at runtime. Either with our own event header with a field type identifying the btf type id, or with some already existing field from perf ring buffers.

Reply:
> In my opinion, the semantics of `macro: WITH_PATHS` are not good: your yaml snippet is presumably gadget.yaml, which is run by `ig run` to load the gadget correctly. But C macros are used at compiled time and I don't think that compilation details should leak in gadget.yaml. The gadget might be built from other languages than C and it should not matter for `ig run`. Maybe it would work better in build.yaml?
> 
You're right. The build.yaml is probably more appropriated for specifying the defines. Using that, we could create two versions of the gadget, but then we are missing the association between each version and a given flag. How can you do that? Any idea?

> But I am still worried about the combinatorial explosion if several fields use such macros.
> 
IMO, that's an issue the gadget developer must take care of. We provide the tools, they are responsible to use them properly.

> I think we should find a way to pass the type in the ring buffer at runtime. Either with our own event header with a field type identifying the btf type id, or with some already existing field from perf ring buffers.
>
Could you please elaborate this idea?

Reply:
I don't have more precise idea, and I don't know if it could work. I just see in the manpage of perf_event_open that it has a lot of options. I wonder if metadata could be useful.

Reply:
I made some experiments to check if we can support optional fields in a more efficient way (without wasting a lot of memory and sending a lot of bytes that aren't used on the ring buffer). I haven't reached any conclusion yet, but want to mention it so you're aware.

TBH I don't think the previous "macro-based" approach will work with image based gadgets, some limitations I see are:
- There is a macro for each optional parameter, the combination explodes (already mentioned by Alban before)
- We need to compile and store two versions of the program in the OCI image (we don't have anything similar yet)
  - We need to understand which version to load based on the parameter
- Testing is more complicated (in the end, those are two different ebpf objects we need to test)

### Idea 

I explored using a Type-length-value encoding approach. Thea idea is to make the gadget send an event that has a variable length, the first bytes will be the fields that are always present (pid, tid, uid, etc), and after it, to have optional fields encoded using TLV

```c
struct event {
 u32 pid;
 u32 tid;
 char optional_fields[0];
}
```

The optional fields will encoded as

```
+---------------+-----------+--------------+---------------+-----------+--------------+
| field0 length | field0 id | field0 value | field1 length | field1 id | field1 value | ...
+---------------+-----------+--------------+---------------+-----------+--------------+
```

The gadget would mark optional fields by using something like:

```c
// execs: name of the tracer 
// cwd: name of the field 
// 1: id of the field 

// TODO: define type of the field (string vs bytes[])

GADGET_TRACER_OPTIONAL_FIELD(execs, cwd, 1)
```

The ebpf operator will then use these macros to add and set the fields when reading the value from the ring buffer. 

Something I haven't explored is how to enable / disable those fields. We already have a [IsRequested()](https://github.com/inspektor-gadget/inspektor-gadget/blob/84f57b00c0e1f5c00e1ed75432b0f468e85da3ed/pkg/datasource/datasource.go#L169) method on the field interface, I think we can wire it to set the value of a constant in the eBPF code when a given field is enabled / disabled.

As mentioned before, it's still some quick experimentation I did, but I think I can make it work. 

Btw, I don't expect to block this PR, please keep it rolling without waiting for this support as it'll take some more time.

Reply:
I think TLV encoding is good. That would also help reducing the size for strings that typically are shorter than MAX_STRING_SIZE (4096). The `cwd` field is typically 2 bytes ("/" with a NULL).

GADGET_TRACER_OPTIONAL_FIELD should include the BTF type too, so that types like `gadget_l3endpoint_t` can be enriched later on.

Reply:
> GADGET_TRACER_OPTIONAL_FIELD should include the BTF type too, so that types like gadget_l3endpoint_t can be enriched later on.

I think we can initially limit it to be only strings and byte arrays. (it seems to me they are the most common use cases). I fear including BTF information here will be complicated.